### PR TITLE
Soft 333 add mpxe pca9539r

### PR DESCRIPTION
**Pull request by JarvisWeng**

### CODE REVIEW

**Review by g-jessmuir — CHANGES_REQUESTED** [review]
I think we need to change it a bit since sending 4kb seems a bit much every message

File: mpxe/protos/stores.proto [hunk 1]
```diff
@@ -4,8 +4,8 @@ enum MxStoreType {
   GPIO = 0;
   ADS1015 = 1;
   MCP2515 = 2;
-  ADS1259 = 3;
```
[the remark this answers is not in the document]
Comment (JarvisWeng) [reply mpxe/protos/stores.proto:7]: oh, i thought i didn't. My bad.

File: libraries/ms-drivers/src/x86/pca9539r_gpio_expander.c [hunk 2]
```diff
@@ -19,11 +79,17 @@ StatusCode pca9539r_gpio_init(const I2CPort i2c_port, const I2CAddress i2c_addre
...
+#ifdef MPXE
+  prv_export();
+#endif
   return STATUS_CODE_OK;
 }
 
 StatusCode pca9539r_gpio_init_pin(const Pca9539rGpioAddress *address,
                                   const Pca9539rGpioSettings *settings) {
+#ifdef MPXE
+  prv_init_store();
```
[the remark this answers is not in the document]
Comment (JarvisWeng) [reply libraries/ms-drivers/src/x86/pca9539r_gpio_expander.c:91]: sure

File: libraries/ms-drivers/src/x86/pca9539r_gpio_expander.c [hunk 3]
```diff
@@ -8,7 +19,56 @@ static I2CPort s_i2c_port = NUM_I2C_PORTS;
...
+  MxPca9539rStore *mask = mx_pca9539r_store__unpack(NULL, mask_buf.len, mask_buf.data);
+
+  for (uint16_t i = 0; i < mask->n_state; i++) {
+    // only update state if mask is set
+    if (mask->state[i] != 0) {
+      s_store.state[i] = msg->state[i];
+      uint8_t i2c_port = i / MAX_I2C_ADDRESSES;
+      uint8_t gpio_pin = i % MAX_I2C_ADDRESSES;
+      if (s_pin_settings[i2c_port][gpio_pin].state != (uint8_t)msg->state[i]) {
+       s_pin_settings[i2c_port][gpio_pin].state = msg->state[i];
```
[the remark this answers is not in the document]
Comment (JarvisWeng) [reply libraries/ms-drivers/src/x86/pca9539r_gpio_expander.c:43]: i have no idea what happened to formatting everywhere

File: libraries/ms-drivers/src/x86/pca9539r_gpio_expander.c [hunk 4]
```diff
@@ -8,7 +19,56 @@ static I2CPort s_i2c_port = NUM_I2C_PORTS;
...
+static void prv_init_store(void) {
+  store_config();
+  StoreFuncs funcs = {
+    (GetPackedSizeFunc)mx_pca9539r_store__get_packed_size,
+    (PackFunc)mx_pca9539r_store__pack,
+    (UnpackFunc)mx_pca9539r_store__unpack,
+    (FreeUnpackedFunc)mx_pca9539r_store__free_unpacked,
+    (UpdateStoreFunc)update_store,
+  };
+  s_store.n_state = MAX_I2C_ADDRESSES * NUM_PCA9539R_GPIO_PINS;
```
[the remark this answers is not in the document]
Comment (JarvisWeng) [reply libraries/ms-drivers/src/x86/pca9539r_gpio_expander.c:62]: sure

File: mpxe/sims/pca9539r.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+class Pca9539r(sim.Sim):
+    def handle_update(self, pm, proj):
+        stores  = proj.stores
+        if (stores_pb2.MxStoreType.PCA9539R, 0) in stores:
+            pca         = stores[(stores_pb2.MxStoreType.PCA9539R, 0)]
+            self.state  = [pca.state[i] for i in range((0x74 * 16), (0x75 * 16))]
+
+    def assert_store_values(self, proj, state):
+        for x in range(16):
+            assert(self.state[x]  == state)
```
[the remark this answers is not in the document]
Comment (JarvisWeng) [reply mpxe/sims/pca9539r.py:15]: Actually this is in the `sim` module, so how would I access the `TestPca9539r`? I only pass in the `proj`, not sure how to go up a layer to get the test

[the remark this answers is not in the document]
Comment (g-jessmuir) [reply mpxe/sims/pca9539r.py:15]: oh wait you’re right, sorry why do we need the assert at all then?

[the remark this answers is not in the document]
Comment (JarvisWeng) [reply mpxe/sims/pca9539r.py:15]: nvm, i figured it out. I thought I had inherit this function.

[the remark this answers is not in the document]
Comment (JarvisWeng) [reply mpxe/sims/pca9539r.py]: Oh, I just assert to make sure the values are being passed to the store correctly. It's ok, it's my function so I passed in the test instead.

**Review by g-jessmuir — APPROVED** [review]

**Pull request merged**